File: M6/testcontainers-vitest-postgresql/test/2.external-sqlfile.test.ts
import { expect, test, beforeAll, afterAll } from 'vitest';
import pg, { Client } from 'pg';
import { PostgreSqlContainer, StartedPostgreSqlContainer } from '@testcontainers/postgresql';
import path from 'path';

let client: typeof Client;
let container: StartedPostgreSqlContainer;

beforeAll(async () => {
  const sqlFileHostPath = path.resolve(__dirname, 'data.sql');
  const sqlFileContainerPath = '/docker-entrypoint-initdb.d/01-init.sql';

  container = await new PostgreSqlContainer('postgres:15-alpine')
    .withCopyFilesToContainer([
      { source: sqlFileHostPath, target: sqlFileContainerPath }
    ])
    .start();

  client = new Client({
    host: container.getHost(),
    port: container.getPort(),
    user: container.getUsername(),
    password: container.getPassword(),
    database: container.getDatabase(),
  });
  await client.connect();
});

afterAll(async () => {
  await client.end();
  await container.stop();
});

// Test 2: fetch all original mock persons
test('should fetch all mock persons from database (external SQL file)', async () => {
  const res = await client.query('SELECT * FROM persons ORDER BY id');
  expect(res.rowCount).toBe(10);
  const names = res.rows.map(r => r.name);
  expect(names).toEqual(['Alice','Bob','Charlie','Diana','Eve','Frank','Grace','Hank','Ivy','Jack']);
});

// Test 3: filter persons older than 30
test('should fetch persons older than 30 (external SQL file)', async () => {
  const res = await client.query('SELECT * FROM persons WHERE age > $1 ORDER BY id', [30]);
  const expectedIds = [4,6,8];
  expect(res.rows.map(r => r.id)).toEqual(expectedIds);
});

File: M6/testcontainers-python-postgresql/tests/2_external_sqlfile_test.py
import os
import pytest
import psycopg2
import psycopg2.extras
from testcontainers.postgres import PostgresContainer
from .db_client import DBClient

MOCK_DATA = [
    {'id': 1, 'name': 'Alice', 'age': 30},
    {'id': 2, 'name': 'Bob', 'age': 24},
    {'id': 3, 'name': 'Charlie', 'age': 28},
    {'id': 4, 'name': 'Diana', 'age': 35},
    {'id': 5, 'name': 'Eve', 'age': 22},
    {'id': 6, 'name': 'Frank', 'age': 33},
    {'id': 7, 'name': 'Grace', 'age': 27},
    {'id': 8, 'name': 'Hank', 'age': 31},
    {'id': 9, 'name': 'Ivy', 'age': 29},
    {'id': 10, 'name': 'Jack', 'age': 26},
]

@pytest.fixture(scope="module")
def db_connection():
    sql_script_path = os.path.join(os.path.dirname(__file__), 'data.sql')
    
    with PostgresContainer("postgres:15-alpine") \
            .with_volume_mapping(os.path.dirname(sql_script_path), "/docker-entrypoint-initdb.d") as postgres:
        
        dsn = postgres.get_connection_url().replace("+psycopg2", "")
        conn = psycopg2.connect(dsn)
        db_client = DBClient(dsn)
        db_client.conn = conn # Assign the established connection
        yield db_client
        db_client.close()
        # use tenacity/retry for proper error handling
        # pytest.fail("Could not connect to the database")
        
def test_should_fetch_all_mock_persons_from_database(db_connection):
    result_dicts = db_connection.fetch_all("SELECT * FROM persons ORDER BY id")

    assert len(result_dicts) == len(MOCK_DATA)
    assert result_dicts == MOCK_DATA


def test_should_fetch_persons_older_than_30(db_connection):
    result_dicts = db_connection.fetch_all("SELECT * FROM persons WHERE age > %s ORDER BY id", [30])
        
    expected = [p for p in MOCK_DATA if p['age'] > 30]
    assert result_dicts == expected
